### DISCUSSION TITLE
VSCode project does not show Android Emulators

### DISCUSSION THREAD

**Discussion by mcNets** [unoplatform/uno#12185] [Q&A] [answered]
I've create a new VSCode project following the instructions that I found in the documentation: https://platform.uno/docs/articles/get-started-vscode.html?tabs=windows%2Candroiddebug

Then, I've opened Visual Studio Android Device Manager and I launched a Pixel -5 API 33 emulator.

In visual studio I selected Uno Mobile target and android debug. According to docs it should appear a device selector next to the os selector but nothing happens.

![image](https://user-images.githubusercontent.com/24267381/236574102-2a647570-4394-4b2f-84d9-c32e74534147.png)

Is there any workaround to get the emulators list?

**Comment by spouliot** [comment]
It's possible to have more than one Android SDK installed (in different locations).

With your Android emulator running try the following command **inside** a VSCode terminal.

```bash
adb devices -l
```

This should list all devices, both physicals and (running) emulators - so it should include the one you're already running.

```bash
List of devices attached
emulator-5554          device product:sdk_gphone64_arm64 model:sdk_gphone64_arm64 device:emu64a transport_id:1
```

<img width="990" alt="Screenshot 2023-05-05 at 7 43 36 PM" src="https://user-images.githubusercontent.com/260465/236585613-ffb7932e-263b-409c-9e0c-412dd9fb4d4c.png">

This is the same command that the extension will use to get the list of devices (but there's a bit more work done to match `emulator-5554` to the `Pixel 5 - API 33` name).

However if there's more than one `adb` command available (IOW more than one Android SDK) then it's possible that the one used by the extension is not the same one you used to open the emulator.

If that does not work can you copy/paste the logs from `Uno Platform - Debugging` ? You can find them in the `OUTPUT` section, there's a combobox to select the type of logs.

**Reply by mcNets** [reply]
@spouliot thanks for the answer.

My first attempt to execute adb when the emulator is running, returns an error:

```
➜ adb devices -1
adb: The term 'adb' is not recognized as a name of a cmdlet, function, script file, or executable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
```

And the Output window show this message:
```
[Info] Starting Debug adapter
[Warn] ios devices cannot be detected on win32
[Info] Forcing project reload to switch to net7.0-ios | Debug
[Info] Forcing project reload to switch to net7.0-android | Debug
[Warn] ios devices cannot be detected on win32
[Info] Forcing project reload to switch to net7.0-ios | Debug
[Info] Forcing project reload to switch to net7.0-android | Debug
```

I don't have WSA installed, is it required?

I was trying to use same emulators that I have installed in Visual Studio.

uno-check give me this results:

```
» Synchronizing configuration... ok
» Scheduling appointments... ok

> OpenJDK 11.0 Checkup...
  - 11.0.16.1 (C:\Program Files\Microsoft\jdk-11.0.16.101-hotspot\bin\..)

> Visual Studio 17.4.4 Checkup...
  - 17.5.5 - C:\Program Files\Microsoft Visual Studio\2022\Professional

> Visual Studio Workloads Checkup...
  - Universal Windows Platform development is installed (17.5.5)
  - Mobile development with .NET is installed (17.5.5)
  - ASP.NET and web development is installed (17.5.5)

> Android SDK Checkup...
  - emulator (31.3.14)
  - build-tools;32.0.0 (32.0.0)
  - platforms;android-32 (1)
  - system-images;android-31;google_apis;x86_64 (8)
  - platform-tools (33.0.2)
  - cmdline-tools;5.0 (5.0)

> Android Emulator Checkup...
  - Emulator: pixel_2_pie_9_0_-_api_28 found.

> .NET SDK Checkup...
  - 3.1.426 - C:\Program Files\dotnet\sdk\3.1.426
  - 5.0.408 - C:\Program Files\dotnet\sdk\5.0.408
  - 6.0.302 - C:\Program Files\dotnet\sdk\6.0.302
  - 6.0.408 - C:\Program Files\dotnet\sdk\6.0.408
  - 7.0.203 - C:\Program Files\dotnet\sdk\7.0.203

> .NET SDK - Workload Deduplication Checkup...

> Edge WebView2 Checkup...
  - Found Edge WebView2 version 112.0.1722.68

> Windows Subsystem for Linux Checkup...
  - Ubuntu (Default)
  - docker-desktop
  - docker-desktop-data

> Windows Python Installation Checkup Checkup...
  - Python is installed in C:\WINDOWS\.

> Windows Long Path Checkup Checkup...
  - Long paths are enabled on Windows!

> Windows Hyper-V Checkup Checkup...
  - Hyper-V is configured

> .NET SDK - Workloads (7.0.102) Checkup...
  - android (Microsoft.NET.Sdk.Android.Manifest-7.0.100 : 33.0.26/7.0.100) installed.
  - ios (Microsoft.NET.Sdk.iOS.Manifest-7.0.100 : 16.1.1481/7.0.100) installed.
  - maccatalyst (Microsoft.NET.Sdk.MacCatalyst.Manifest-7.0.100 : 16.1.1481/7.0.100) installed.
  - macos (Microsoft.NET.Sdk.macOS.Manifest-7.0.100 : 13.0.2036/7.0.100) installed.
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

- Congratulations, everything looks great!
```

I've added the path to the adb.exe tool in the environment variables, now it can be executed, but still doesn't work either.

**Reply by spouliot** [reply]
> My first attempt to execute adb when the emulator is running, returns an error:

ok, that explains why a list of devices cannot be shown.

> I don't have WSA installed, is it required?

No, it's entirely optional.

> uno-check give me this results:

Must be missing a location check.... I'll double check every locations that `uno-check` probes for the Android SDK. The extension already checks 3 different places (on Windows) to look for `adb` but it might be missing one (or more). 

> I've added the path to the adb.exe tool in the environment variables, now it can be executed, but still doesn't work either.

Can it be executed inside VSCode terminal ? 

If not make sure that the directory is visible when issuing a `PATH` command inside a Command Prompt (inside VSCode). 

Depending on how you have set the variable and how you start VSCode it's possible that the VSCode process does not have the updated PATH variable yet.

**Reply by mcNets** [reply]
>Can it be executed inside VSCode terminal ?

Yes, it can be executed inside VSCode terminal:

➜ adb devices -l
List of devices attached
emulator-5554          device product:sdk_gphone64_x86_64 model:sdk_gphone64_x86_64 device:emu64x transport_id:2

But still can't see/select any device executing the command: 'Uno Platform: Select the device for execution.'

I've tried in Visual Studio 2022 17.5.5 and I can run a Uno app in the Pixel emulator

**Reply by spouliot** [reply]
Ok, so it's in the PATH of the terminal.

Let's go from simple to more complex things to try.

1. if you have no already, restart VSCode. In case the current process might be older than the environment variable change.

2. inside a VSCode terminal, type `set` and share the output. Please sanitize the data as some variables could be sensitive. I'm looking for `ANDROID_*` variables that could override the default (PATH) SDK.

3. inside a VSCode terminal type:

```bash
dotnet c:\Users\sebas\.vscode\extensions\unoplatform.vscode-0.9.4\bin\dwatcher\net6.0\dwatcher.dll . android -o
```

note: change `sebas` for your user name

Normally (for success) there would not be any output on the console. Please paste here any error message. 

Then type:

```bash
type uno.android.json
```

and this should contains some JSON like:

```json
[{"identifier":"","platform":"com.google.platform.androidemulator","architecture":"arm64","modelName":"Android_Emulator_31","operatingSystemVersion":"android-31","interface":"","name":"Android_Emulator_31"},{"identifier":"127.0.0.1:58526","platform":"com.google.platform.android","architecture":"arm64","modelName":"Microsoft Corporation Subsystem for Android(TM)","operatingSystemVersion":"android-33","interface":"","name":"127.0.0.1:58526"},{"identifier":"","platform":"com.google.platform.androidemulator","architecture":"x86_64","modelName":"Pixel_3a_API_33_x86_64","operatingSystemVersion":"android-33","interface":"","name":"Pixel_3a_API_33_x86_64"}]
```

Your file (if it exists) is likely empty.

**Reply by mcNets** [reply]
>inside a VSCode terminal, type set and share the output. Please sanitize the data as some variables could be sensitive. I'm looking for ANDROID_* variables that could override the default (PATH) SDK.

C:\Program Files (x86)\Android\android-sdk\platform-tools;

(That's the path I manually added to the environment variables where you can find adb.exe)

>dotnet c:\Users\sebas\.vscode\extensions\unoplatform.vscode-0.9.4\bin\dwatcher\net6.0\dwatcher.dll . android -o

returns nothing.

That's the content of uno.android.json file:

```
[{"identifier":"","platform":"com.google.platform.androidemulator","architecture":"x86_64","modelName":"Pixel 5 - API 30","operatingSystemVersion":"android-30","interface":"","name":"pixel_5_-_api_30"},{"identifier":"","platform":"com.google.platform.androidemulator","architecture":"x86_64","modelName":"Pixel 5 - API 33","operatingSystemVersion":"android-33","interface":"","name":"pixel_5_-_api_33"}]
```

**Reply by spouliot** [reply]
I did not expect all steps to work 😕 

Beside `PATH` what are the name(s) of the environment variable(s) pointing to `C:\Program Files (x86)\Android\android-sdk\platform-tools;` ?

Also can you confirm that you have the same `uno.android.json` file under `c:\Users\sebas\AppData\Roaming\Code\User\globalStorage\unoplatform.vscode\` ?

note: change `sebas` for your user name

This is where the extension ask the file to be generated (and from where it's loaded).

**Reply by mcNets** [reply]
>Beside PATH what are the name(s) of the environment variable(s) pointing to C:\Program Files (x86)\Android\android-sdk\platform-tools; ?

None

>Also can you confirm that you have the same uno.android.json file under c:\Users\sebas\AppData\Roaming\Code\User\globalStorage\unoplatform.vscode\ ?

There is no unoplatform.vscode folder there.
I have one **uno.android.json** file in my project's folder.

**Reply by spouliot** [reply]
> None

ok, that's a misunderstanding.

Inside a VSCode terminal, type `set` and share the output.

You can sanitize the output before sharing the content (as some data could be sensitive).

I'm looking for variables that could confuse (e.g. override) the logic that checks for the Android SDK.

> There is no unoplatform.vscode folder there.

That's unexpected. Are you able to create that directory ? (it might be a permission issue)

If creating the directory works then restart VSCode and try again.

**Reply by mcNets** [reply]
>Inside a VSCode terminal, type set and share the output.

```
ALLUSERSPROFILE=C:\ProgramData
APPDATA=C:\Users\Joan Magnet\AppData\Roaming
CommonProgramFiles=C:\Program Files\Common Files
CommonProgramFiles(x86)=C:\Program Files (x86)\Common Files
CommonProgramW6432=C:\Program Files\Common Files
ComSpec=C:\WINDOWS\system32\cmd.exe
CONDA_PROMPT_MODIFIER=False
DriverData=C:\Windows\System32\Drivers\DriverData
ESET_OPTIONS=                                                                                                                                                                                                                                                                                                                                                                                                           

HOMEDRIVE=C:
HOMEPATH=\Users\Joan Magnet
JAVA_HOME=C:\Program Files\Eclipse Foundation\jdk-8.0.302.8-hotspot
LOCALAPPDATA=C:\Users\Joan Magnet\AppData\Local
NUMBER_OF_PROCESSORS=12
ORIGINAL_XDG_CURRENT_DESKTOP=undefined
OS=Windows_NT
Path=C:\Users\Joan Magnet\.oh-my-posh;C:\Program Files\PowerShell\7;C:\Program Files (x86)\Common Files\Oracle\Java\javapath;C:\Program Files\Microsoft\jdk-11.0.16.101-hotspot\bin;C:\Program Files\Python39\Scripts\;C:\Program Files\Python39\;C:\Program Files\Python310\Scripts\;C:\Program Files\Python310\;C:\Program Files\Eclipse Foundation\jdk-8.0.302.8-hotspot\bin;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Windows\System32\OpenSSH\;C:\Program Files\Intel\WiFi\bin\;C:\Program Files\Common Files\Intel\WirelessCommon\;C:\Program Files (x86)\NVIDIA Corporation\PhysX\Common;C:\Program Files\Microsoft SQL Server\130\Tools\Binn\;C:\Program Files\Microsoft SQL Server\Client SDK\ODBC\170\Tools\Binn\;C:\Program Files (x86)\Common Files\Acronis\VirtualFile\;C:\Program Files (x86)\Common Files\Acronis\VirtualFile64\;C:\Program Files (x86)\Common Files\Acronis\FileProtector\;C:\Program Files (x86)\Common Files\Acronis\FileProtector64\;C:\Program Files (x86)\Common Files\Acronis\SnapAPI\;C:\Program Files\Microsoft SQL Server\150\Tools\Binn\;C:\Program Files\NVIDIA Corporation\NVIDIA NvDLISR;C:\Program Files (x86)\Microsoft SQL Server\150\Tools\Binn\;C:\Program Files\Microsoft SQL Server\150\DTS\Binn\;C:\Program Files (x86)\Microsoft SQL Server\150\DTS\Binn\;C:\Program Files\Azure Data Studio\bin;C:\Program Files\dotnet\;C:\Program Files\GTK3-Runtime Win64\bin;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\WINDOWS\System32\WindowsPowerShell\v1.0\;C:\WINDOWS\System32\OpenSSH\;C:\Program Files\Docker\Docker\resources\bin;C:\Program Files\Git\cmd;C:\Program Files\PowerShell\7\;C:\Program Files (x86)\Android\android-sdk\platform-tools;C:\Users\Joan Magnet\AppData\Local\Microsoft\WindowsApps;C:\Users\Joan Magnet\AppData\Local\Programs\Azure Data Studio\bin;C:\Users\Joan Magnet\.dotnet\tools;C:\Users\Joan Magnet\AppData\Local\Programs\Fiddler;C:\Users\Joan Magnet\.dotnet\tools;C:\Users\Joan Magnet\AppData\Local\Microsoft\WindowsApps;C:\Users\Joan Magnet\AppData\Local\Programs\Microsoft VS Code\bin;C:\Users\Joan Magnet\AppData\Local\Programs\oh-my-posh\bin
PATHEXT=.COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC;.PY;.PYW;.CPL
POSH_GIT_ENABLED=False
POSH_PATH=C:\Users\Joan Magnet\.oh-my-posh
POSH_THEME=C:\Users\Joan Magnet\.oh-my-posh\themes\craver.omp.json
POSH_THEMES_PATH=C:\Users\Joan Magnet\.oh-my-posh\themes
POWERLINE_COMMAND=oh-my-posh
POWERSHELL_DISTRIBUTION_CHANNEL=MSI:Windows 10 Pro
PROCESSOR_ARCHITECTURE=AMD64
PROCESSOR_IDENTIFIER=Intel64 Family 6 Model 166 Stepping 0, GenuineIntel
PROCESSOR_LEVEL=6
PROCESSOR_REVISION=a600
ProgramData=C:\ProgramData
ProgramFiles=C:\Program Files
ProgramFiles(x86)=C:\Program Files (x86)
ProgramW6432=C:\Program Files
PROMPT=$P$G
PSModulePath=C:\Users\Joan Magnet\Documents\PowerShell\Modules;C:\Program Files\PowerShell\Modules;c:\program files\powershell\7\Modules;C:\Program Files\WindowsPowerShell\Modules;C:\WINDOWS\system32\WindowsPowerShell\v1.0\Modules;C:\Program Files (x86)\Microsoft SQL Server\150\Tools\PowerShell\Modules\
PUBLIC=C:\Users\Public
SystemDrive=C:
SystemRoot=C:\WINDOWS
TEMP=C:\Users\JOANMA~1\AppData\Local\Temp
TMP=C:\Users\JOANMA~1\AppData\Local\Temp
USERNAME=Joan Magnet
USERPROFILE=C:\Users\Joan Magnet
windir=C:\WINDOWS
WSLENV=WT_SESSION::WT_PROFILE_ID
WT_PROFILE_ID={5fb123f1-af88-5b5c-8953-d14a8def1978}
WT_SESSION=7bfb5e56-cd8e-4c26-a263-9da3896e6ff5
ZES_ENABLE_SYSMAN=1
TERM_PROGRAM=vscode
TERM_PROGRAM_VERSION=1.78.0
LANG=en_US.UTF-8
COLORTERM=truecolor
```

>That's unexpected. Are you able to create that directory ? (it might be a permission issue)

I've created the directory, but nothing happens.
Should I add some file there?

Is there any workaround by uninstalling and re-installing some extension?

**Reply by spouliot** [reply]
> Should I add some file there?

You can put the `uno.android.json` file that you manually generated inside the directory. That might work...

I _think_ I have found the issue. I'm testing this and will give you a link to a newer build. It will also include more details in the logs, so it will be helpful even if it does not fix it.

> Is there any workaround by uninstalling and re-installing some extension?

It's unlikely to workaround the issue.

When you uninstall the Uno extension the directory should be deleted. It's a global storage location provided by VSCode (and it should cleanup after itself).

When you re-install the Uno extension the directory should be created when emulators (Android) or simulators (iOS) are found (it's the extension job to create the directory when missing).

**Reply by spouliot** [reply]
A new version should be available soon (once validated) and should fix the issue you have been facing. Thanks for your help to diagnose the issue!

**Reply by mcNets** [reply]
Thanks for your help @spouliot .

**Reply by jeromelaban** [reply, marked as answer]
The build is being published (0.9.5), it should be available shortly in vscode's updates.

**Reply by mcNets** [reply]
Now it works. 

![image](https://github.com/unoplatform/uno/assets/24267381/cb35e42e-cc2a-4f07-b071-ce1e2bf7f384)

**Reply by jeromelaban** [reply]
Thanks for the update!

**Comment by spouliot** [comment]
> Is there any workaround to get the emulators list?

You can also use the command palette (`F1`) and select the following command

<img width="605" alt="Screenshot 2023-05-05 at 7 50 42 PM" src="https://user-images.githubusercontent.com/260465/236586019-1b7c1fbe-9087-4bf2-9900-b5e7813a56b0.png">

This is identical to clicking on the (normally visible) device name in the status bar.

This will bring the list of all emulators (for the Android SDK) and all physically connected Android devices.

<img width="604" alt="Screenshot 2023-05-05 at 7 50 57 PM" src="https://user-images.githubusercontent.com/260465/236586022-c806ebe2-f88a-447c-bfeb-24aae3e16909.png">

If running the icon will have a small _play_ button.
